### ISSUE TITLE
Analytics setup not recognizing an existing UA or GA4 tag - if added via a manually inserted GTM container 

### ISSUE THREAD

**Issue by jamesozzie** [google/site-kit-wp#4103]
## Bug Description

If a user has Tag Manager implemented on their site via a manual snippet insertion (and not via SK) Analytics within Site Kit will not recognize any existing UA code firing from that GTM container. This results in Site Kit inserting the Analytics snippet on a users site by default, with possible skewed data. 

This isn't the case if a user implements GTM (with a UA tag) via Site Kit. 

## Steps to reproduce

1. Create a GTM container that includes a Universal Analytics tag
2. Insert the GTM snippet on a WordPress site manually - into a themes templates
3. Ensure the tag is firing using tagassistant.google.com
4. Connect the Analytics module within Site Kit
5. The plugin does not detect an existing Universal Analytics snippet

## Screenshots

_Tag setup with Google Tag Manager below. Published contained with a UA Analytics snippet_

![image](https://user-images.githubusercontent.com/41326532/134321907-37b0e2bf-cc0c-49ae-8bad-d663c6d97672.png)

_The GTM snippet manually added into a themes header.php file_

![image](https://user-images.githubusercontent.com/41326532/134322528-6a0dd7db-c02b-44ba-b5ab-b4c01ffc19c4.png)

_Confirmation using tagassistant.google.com that the snippet is working as expected below_

![image](https://user-images.githubusercontent.com/41326532/134322422-d769a468-ce19-4606-b805-89ea12922bb7.png)

_Analytics setup experience within Site Kit - the plugin doesn't determine an Analytics snippet contained with a GTM container (only when this container snippet is manually inserted into a themes templates)_ ([Video example](https://recordit.co/MVzIEjIzEb))

![image](https://user-images.githubusercontent.com/41326532/134323399-eabfc7c2-a306-453a-a39b-e578f7350cbf.png)

## Additional Context

 - Site Kit 1.41.0
 - This does not occur with a Site Kit inserted GTM container
 - When a Universal Tag is added to a GTM container the actual Analytics property ID is not printed and visible in a sites source code (This is also the case if the contained was added via Site Kit)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* During the Analytics setup/settings logic, if Tag Manager is active, Site Kit should also check if the Tag Manager container contains a Google Analytics 4 measurement ID, similar to how it already does for a Universal Analytics property ID.
    * A good approach might be to introduce a selector `getSingleGA4MeasurementID` in the `modules/tagmanager` data store (similar to the existing `getSingleAnalyticsPropertyID` which is for UA - technically we could probably rename that then to `getSingleUAPropertyID`).
* If there is an existing GA4 tag, that should also be incorporated into the existing GTM tag notice. We shouldn't incorporate it into the error though - while that is inconsistent for now, we will soon remove that entire error state entirely, meaning it's already as of now considered not critical if the site has multiple tags.
* What should be prevented is that a site has multiple tags of the same property: So if the currently selected GA4 property is the same like the one from the existing GTM tag, the GA4 `useSnippet` setting should be set to `false`.

## Implementation Brief
* Using `assets/js/modules/tagmanager/datastore/versions.js`,
	* Add a new selector `getGA4MeasurementIDs` which should largely follow what we have for the `getAnalyticsPropertyIDs` selector with the following changes:
		* The AMP-related code should **not** be copied, AMP not yet supporting GA4.
		* Instead of calling the `getLiveContainerAnalyticsPropertyID` selector, create another selector `getLiveContainerGA4MeasurementID` which has the same parameters as `getLiveContainerAnalyticsPropertyID` and use almost the same code with the following changes:
			* `getLiveContainerAnalyticsTag` should be replaced by another utility function called `getLiveContainerGA4Tag` which looks for a tag with type `gaawc` (Google Analytics 4 Configuration) instead of `ua` / `ua_amp`.
			* Check if there is a `measurementId` key instead of `trackingId` in `analyticsTag.parameter`.
			* Remove the check for/references to `gaSettings`, as `gaSettings` is not used in GA4 tags.
			* Replace `trackingId` with `value` in the variable lookup.
			* `isValidPropertyID` should be replaced by another utility function called `isValidGA4MeasurementID` which checks whether the `measurementId` is a valid Google Analytics Measurement ID, i.e starting with `G-`.
* Using `assets/js/modules/analytics/components/common/ExistingGTMPropertyNotice.js`,
	* Query the TagManager data store via the `getGA4MeasurementIDs` selector. If there the return value is not an empty array, incorporate the GA4 measurement IDs into the warning message and show it.
* Using `assets/js/modules/analytics/components/setup/SetupFormGA4Transitional.js`:
	* Render the `ExistingGTMPropertyNotice` if there is no existing UA or GA4 tag.
* Using `assets/js/modules/analytics/components/setup/SetupFormGA4.js`:
	* Render the `ExistingGTMPropertyNotice` if there is no existing GA4 tag.
* Using `assets/js/modules/analytics/components/setup/SetupForm.js`,
	* Query the selected measurement ID via the `getMeasurementID` selector of the `analytics-4` data store. If the measurement ID exists in the results from `getGA4MeasurementIDs`, then dispatch the `setUseSnippet` action of the `analytics-4` data store, with `false` as parameter.
* Using `assets/js/modules/analytics/components/common/GA4ActivateSwitch.js`, don't display the GA4 toggle if the conditions in the previous bullet point are met.

### Test Coverage
* New tests to be added for new selectors.

## QA Brief

* 

## Changelog entry

*

**Comment by jamesozzie** [comment]
As a follow up I also performed a check with GA4 added to the container - the container which has it's GTM snippet manually inserted into the themes files. The Analytics setup within Site Kit will also not recognize a GA4 snippet in the above scenario. 

_GA4 added via GTM - confirmed to be working and firing_ 

![image](https://user-images.githubusercontent.com/41326532/134329984-5296e93f-1e04-47a6-a5a7-c3c0692f8b3a.png)

_Site Kit setup experience - GA4 not recognized when a GA4 property is already firing via GTM_ ([Video demo](https://recordit.co/5fhz0ceoel))

![image](https://user-images.githubusercontent.com/41326532/134330941-56f158ba-7cdc-4845-b2eb-b0d8a97a27f4.png)

**Comment by aaemnnosttv** [comment]
> * Rename the `getSingleAnalyticsPropertyID` selector to `getSingleUAPropertyID` and update references across the codebase.

@asvinb This might end up being a bit more than a single renaming as there are multiple selectors in that store slice which should be updated for consistency if we were to do this. 

![image](https://user-images.githubusercontent.com/1621608/140956234-d965c83c-de9b-4a6c-b59d-b4dfa9b51ff0.png)

While I think this is generally a good idea, it might be best to leave this for a follow-up issue as it would otherwise make the PR for this issue much larger which we should try to avoid for this area of the codebase as it's already rather complex.

> * Add a new selector `getAnalyticsMeasurementIDs` which should largely follow what we have for the `getAnalyticsPropertyIDs` selector [...]

As mentioned in the ACs (although not a requirement) let's use `GA4` instead of `Analytics` in the selectors for GA4 for consistency with others.

It's worth noting that the current UA version of this selector is plural because GTM uses different containers in web and AMP contexts. However, AMP does not currently support GA4, nor is it possible to add a GA4 tag to an AMP container. Nonetheless, we should probably ignore this detail right now and pretend it does work for now to avoid extra changes later if/when this support is added. This way we should be able to follow the existing implementation very closely which should help keep this simple. The one caveat as mentioned in the ACs is that as far as the versions store slice is concerned, we don't really care if there are multiple unique measurement IDs, the only concern here is that the GA4 use snippet is disabled if the same measurement ID is detected in the GTM container.

All in all though, the IB LGTM. Thoughts @felixarntz ?

**Comment by felixarntz** [comment]
@asvinb +1 to what @aaemnnosttv said above. Let's use `GA4` in the names (e.g. `getGA4MeasurementIDs`) for all the new selectors etc. I also agree we should be using plural even if right now GA4 isn't supported in AMP and right now it would only be one measurement ID. I don't have a strong opinion on the renaming from `Analytics` to `UA`, but definitely if we do it here, it should apply to all applicable `modules/tagmanager` selectors. If this touches too much of the codebase though, let's keep this for another PR like @aaemnnosttv said.

@asvinb Once you update the naming-related parts in the IB based on the above, this should be good to go.

**Comment by asvinb** [comment]
Thanks @aaemnnosttv @felixarntz , IB has been updated with the new selector names.

**Comment by felixarntz** [comment]
@asvinb Thanks! One final thing I missed before:

> Using `assets/js/modules/analytics/components/common/UseUAandGA4SnippetSwitches.js`, don't display the GA4 toggle is the conditions in the previous bullet point are met.

This component no longer exists as of #3807 - can you update the IB to refer to the new component?

**Comment by asvinb** [comment]
Updated @felixarntz !

**Comment by felixarntz** [comment]
IB ✅

**Comment by techanvil** [comment]
Hi @felixarntz, @asvinb, @aaemnnosttv. While implementing this ticket, I've noticed a number of issues related to the AC and IB, which I have outlined below. I hope this makes sense, please let me know if anything needs to be clarified, or if I've got something wrong. Happy to jump on a call to discuss.

* The bug describes UA and then goes on to mention GA4. The AC, possibly in reference to the related ticket #3998, assumes that the UA flow is OK and only focuses on the GA4 solution. However in fact the bug mentioned is present for both UA and GA4 in the transitional `SetupFormGA4Transitional` setup flow, which is not addressed in the AC/IB.
* The AC/IB doesn't cover preselecting and disabling the GA4 dropdown in the Analytics flow when a GT4 tag exists, in the same way the UA dropdown is preselected/disabled when a UA tag exists. Additionally the UA dropdown is **not** currently preselected/disabled on the `SetupFormGA4Transitional` flow when a UA tag exists.
* There is an assumption that `getLiveContainerAnalyticsPropertyID` and the call stack it invokes will translate to `getLiveContainerGA4MeasurementID` etc with minimal changes. However this is not as described in the IB.
  * GTM AMP containers don't support GA4 so all the AMP related code should be removed, there's no way of knowing it would work if AMP eventually does support GA4. There's a bit of related discussion here https://github.com/ampproject/amphtml/issues/24621.
  * GA4 tag responses are different to UA beyond what is mentioned in the IB.
    * The current code looks for a tag type of `ua` / `ua_web`. For GA4 the tag type can be `gaawc` for Google Analytics 4 Configuration tag or `gaawe` for Google Analytics 4 Event. It's not clear if we should cover just one or both of these.
    * The variable expansion is also different and `gaSettings` variables can't be used for GA4 as they are UA-specific expecting a UA-formatted tracking ID. It's possible to use a `constant` variable type to specify a GA4 measurement ID. It's not clear if other variables should be used. The current code expects a `gaSettings` variable and this should be removed and possibly replaced with alternative variable expansion.
    * See the example `live-container-version` response below.
* The IB specifies changes to [`assets/js/modules/analytics/components/common/ExistingTagNotice.js`](https://github.com/google/site-kit-wp/blob/1735da6226a671f86b1e043796ec4710729e7e06/assets/js/modules/analytics/components/common/ExistingTagNotice.js) but that seems at odds with the fact the current UA message and GTM-related text is in [`assets/js/modules/analytics/components/common/ExistingGTMPropertyNotice.js`](https://github.com/google/site-kit-wp/blob/1735da6226a671f86b1e043796ec4710729e7e06/assets/js/modules/analytics/components/common/ExistingGTMPropertyNotice.js). It looks like it would be more appropriate to add to `ExistingGTMPropertyNotice` and reuse this in the GA4-related flows, i.e. `SetupFormGA4` and `SetupFormGA4Transitional`.
https://github.com/google/site-kit-wp/blob/1735da6226a671f86b1e043796ec4710729e7e06/assets/js/modules/analytics/components/setup/SetupForm.js#L97-L102
* The IB mentions showing the same message we currently have when `getAnalyticsPropertyIDs` returns a non-empty array, but all the existing copy is designed with expectation of a single property/measurement ID, the IB doesn't cover what happens when `getAnalyticsPropertyIDs` returns more than one ID.

<details>
<summary>
Here's an example `live-container-version` response containing GA4 tags which refer to `constant` variables, as well as a UA tag which uses `gaSettings`:
</summary>

```json
{
  "accountId": "6005157207",
  "containerId": "53572005",
  "containerVersionId": "12",
  "deleted": null,
  "description": null,
  "fingerprint": "1638963755443",
  "name": "change name",
  "path": "accounts/6005157207/containers/53572005/versions/12",
  "tagManagerUrl": "https://tagmanager.google.com/#/versions/accounts/6005157207/containers/53572005/versions/12?apiLink=version",
  "container": {
    "accountId": "6005157207",
    "containerId": "53572005",
    "domainName": null,
    "fingerprint": "1635175212749",
    "name": "foo.sitekit.10uplabs.com",
    "notes": null,
    "path": "accounts/6005157207/containers/53572005",
    "publicId": "GTM-M7Q5V2S",
    "tagManagerUrl": "https://tagmanager.google.com/#/container/accounts/6005157207/containers/53572005/workspaces?apiLink=container",
    "usageContext": [
      "web"
    ]
  },
  "tag": [
    {
      "accountId": "6005157207",
      "blockingRuleId": null,
      "blockingTriggerId": null,
      "containerId": "53572005",
      "fingerprint": "1638963716829",
      "firingRuleId": null,
      "firingTriggerId": [
        "2147479553"
      ],
      "liveOnly": null,
      "monitoringMetadataTagNameKey": null,
      "name": "Foo Page View - UA w/associated",
      "notes": null,
      "parentFolderId": null,
      "path": null,
      "paused": null,
      "scheduleEndMs": null,
      "scheduleStartMs": null,
      "tagFiringOption": "oncePerEvent",
      "tagId": "3",
      "tagManagerUrl": null,
      "type": "ua",
      "workspaceId": null,
      "parameter": [
        {
          "key": "overrideGaSettings",
          "type": "boolean",
          "value": "false"
        },
        {
          "key": "trackType",
          "type": "template",
          "value": "TRACK_PAGEVIEW"
        },
        {
          "key": "gaSettings",
          "type": "template",
          "value": "{{Foo - UA Tracking ID}}"
        }
      ],
      "monitoringMetadata": {
        "key": null,
        "type": "map",
        "value": null
      },
      "consentSettings": {
        "consentStatus": "notSet"
      }
    },
    {
      "accountId": "6005157207",
      "blockingRuleId": null,
      "blockingTriggerId": null,
      "containerId": "53572005",
      "fingerprint": "1638895993160",
      "firingRuleId": null,
      "firingTriggerId": [
        "2147479553"
      ],
      "liveOnly": null,
      "monitoringMetadataTagNameKey": null,
      "name": "Foo Page View - GA4",
      "notes": null,
      "parentFolderId": null,
      "path": null,
      "paused": null,
      "scheduleEndMs": null,
      "scheduleStartMs": null,
      "tagFiringOption": "oncePerEvent",
      "tagId": "6",
      "tagManagerUrl": null,
      "type": "gaawc",
      "workspaceId": null,
      "parameter": [
        {
          "key": "sendPageView",
          "type": "boolean",
          "value": "true"
        },
        {
          "key": "enableSendToServerContainer",
          "type": "boolean",
          "value": "false"
        },
        {
          "key": "measurementId",
          "type": "template",
          "value": "{{Foo - GA4 Measurement ID}}"
        }
      ],
      "monitoringMetadata": {
        "key": null,
        "type": "map",
        "value": null
      },
      "consentSettings": {
        "consentStatus": "notSet"
      }
    },
    {
      "accountId": "6005157207",
      "blockingRuleId": null,
      "blockingTriggerId": null,
      "containerId": "53572005",
      "fingerprint": "1638897179369",
      "firingRuleId": null,
      "firingTriggerId": [
        "2147479553"
      ],
      "liveOnly": null,
      "monitoringMetadataTagNameKey": null,
      "name": "Foo Page View - GA4 Event",
      "notes": null,
      "parentFolderId": null,
      "path": null,
      "paused": null,
      "scheduleEndMs": null,
      "scheduleStartMs": null,
      "tagFiringOption": "oncePerEvent",
      "tagId": "7",
      "tagManagerUrl": null,
      "type": "gaawe",
      "workspaceId": null,
      "parameter": [
        {
          "key": "eventName",
          "type": "template",
          "value": "Test Prop GA4"
        },
        {
          "key": "measurementId",
          "type": "template",
          "value": "none"
        },
        {
          "key": "measurementIdOverride",
          "type": "template",
          "value": "{{Foo - GA4 Measurement ID 2}}"
        }
      ],
      "monitoringMetadata": {
        "key": null,
        "type": "map",
        "value": null
      },
      "consentSettings": {
        "consentStatus": "notSet"
      }
    }
  ],
  "variable": [
    {
      "accountId": "6005157207",
      "containerId": "53572005",
      "disablingTriggerId": null,
      "enablingTriggerId": null,
      "fingerprint": "1638959080010",
      "name": "Foo - UA Tracking ID",
      "notes": null,
      "parentFolderId": null,
      "path": null,
      "scheduleEndMs": null,
      "scheduleStartMs": null,
      "tagManagerUrl": null,
      "type": "gas",
      "variableId": "4",
      "workspaceId": null,
      "parameter": [
        {
          "key": "cookieDomain",
          "type": "template",
          "value": "auto"
        },
        {
          "key": "doubleClick",
          "type": "boolean",
          "value": "false"
        },
        {
          "key": "setTrackerName",
          "type": "boolean",
          "value": "false"
        },
        {
          "key": "useDebugVersion",
          "type": "boolean",
          "value": "false"
        },
        {
          "key": "useHashAutoLink",
          "type": "boolean",
          "value": "false"
        },
        {
          "key": "decorateFormsAutoLink",
          "type": "boolean",
          "value": "false"
        },
        {
          "key": "enableLinkId",
          "type": "boolean",
          "value": "false"
        },
        {
          "key": "enableEcommerce",
          "type": "boolean",
          "value": "false"
        },
        {
          "key": "trackingId",
          "type": "template",
          "value": "UA-214757749-1"
        }
      ]
    },
    {
      "accountId": "6005157207",
      "containerId": "53572005",
      "disablingTriggerId": null,
      "enablingTriggerId": null,
      "fingerprint": "1638895984110",
      "name": "Foo - GA4 Measurement ID",
      "notes": null,
      "parentFolderId": null,
      "path": null,
      "scheduleEndMs": null,
      "scheduleStartMs": null,
      "tagManagerUrl": null,
      "type": "c",
      "variableId": "8",
      "workspaceId": null,
      "parameter": [
        {
          "key": "value",
          "type": "template",
          "value": "G-7TG3MD6C9G"
        }
      ]
    },
    {
      "accountId": "6005157207",
      "containerId": "53572005",
      "disablingTriggerId": null,
      "enablingTriggerId": null,
      "fingerprint": "1638897162501",
      "name": "Foo - GA4 Measurement ID 2",
      "notes": null,
      "parentFolderId": null,
      "path": null,
      "scheduleEndMs": null,
      "scheduleStartMs": null,
      "tagManagerUrl": null,
      "type": "c",
      "variableId": "9",
      "workspaceId": null,
      "parameter": [
        {
          "key": "value",
          "type": "template",
          "value": "G-RL1PM5SSPK"
        }
      ],
      "formatValue": {
        "caseConversionType": null
      }
    }
  ],
  "builtInVariable": [
    {
      "accountId": "6005157207",
      "containerId": "53572005",
      "name": "Page URL",
      "path": null,
      "type": "pageUrl",
      "workspaceId": null
    },
    {
      "accountId": "6005157207",
      "containerId": "53572005",
      "name": "Page Hostname",
      "path": null,
      "type": "pageHostname",
      "workspaceId": null
    },
    {
      "accountId": "6005157207",
      "containerId": "53572005",
      "name": "Page Path",
      "path": null,
      "type": "pagePath",
      "workspaceId": null
    },
    {
      "accountId": "6005157207",
      "containerId": "53572005",
      "name": "Referrer",
      "path": null,
      "type": "referrer",
      "workspaceId": null
    },
    {
      "accountId": "6005157207",
      "containerId": "53572005",
      "name": "Event",
      "path": null,
      "type": "event",
      "workspaceId": null
    },
    {
      "accountId": "6005157207",
      "containerId": "53572005",
      "name": "Container ID",
      "path": null,
      "type": "containerId",
      "workspaceId": null
    }
  ]
}

```
</details>

**Comment by aaemnnosttv** [comment]
Thanks @techanvil I'll try to clarify

> * The bug describes UA and then goes on to mention GA4. The AC, possibly in reference to the related ticket [Site Kit doesn't recognize a Tag Manager placed GA4 snippet #3998](https://github.com/google/site-kit-wp/issues/3998), assumes that the UA flow is OK and only focuses on the GA4 solution. However in fact the bug mentioned is present for both UA and GA4 in the transitional `SetupFormGA4Transitional` setup flow, which is not addressed in the AC/IB.

My understanding of the original issue here is about Analytics being unaware of an existing GTM tag when placed manually **and** Tag Manager is not connected. If the module was connected with the existing tag present then Analytics would also be aware of it but only if the user had access to the container as setup is blocked otherwise. If the user doesn't have access to the container then we cannot inspect the tags it has as this requires the internal container ID which can only be known if you have access to it on the service.

Technically if Tag Manager was active, we could check for an existing tag on the fly and attempt to fetch its tags during Analytics setup, but this is adding even more complexity to something which is already quite complex. The existing tag detection and handling is far from perfect and we actually went a bit too far with it initially.

So this issue became more of a solution for #3998 and in retrospect, perhaps this one should have been closed instead.

> * The AC/IB doesn't cover preselecting and disabling the GA4 dropdown in the Analytics flow when a GT4 tag exists, in the same way the UA dropdown is preselected/disabled when a UA tag exists.

This is intentional as such force-selection was decided against starting with GA4. This is what is referred to in the ACs _"while that is inconsistent for now, we will soon remove that error state entirely, meaning it's already as of now considered not critical if the site has multiple tags."_

> * Additionally the UA dropdown is **not** currently preselected/disabled on the `SetupFormGA4Transitional` flow when a UA tag exists.

This sounds like it may be a bug although it may have been intentional as a step towards the removal of such restrictions. @eugene-manuilov did we change this recently?

> * GTM AMP containers don't support GA4 so all the AMP related code should be removed, there's no way of knowing it would work if AMP eventually does support GA4.

Yes, this was [discussed a bit above](https://github.com/google/site-kit-wp/issues/4103#issuecomment-964497762) but looks like it wasn't mentioned in the IB. We only need to consider the web container right now but essentially the approach regarding selector names was to mimic the ones we have now for UA (i.e. plural) to avoid more substantial changes later if/when AMP ever supports GA4.

> GA4 tag responses are different to UA beyond what is mentioned in the IB.
> * The current code looks for a tag type of `ua` / `ua_web`. For GA4 the tag type can be `gaawc` for Google Analytics 4 Configuration tag or `gaawe` for Google Analytics 4 Event. It's not clear if we should cover just one or both of these.

Google Analytics 4 Event is only for sending additional custom events; it isn't interchangeable with the GA4 configuration tag (`gaawc`) which is where the measurement ID is set.

> * The variable expansion is also different and `gaSettings` variables can't be used for GA4 as they are UA-specific expecting a UA-formatted tracking ID. It's possible to use a `constant` variable type to specify a GA4 measurement ID. It's not clear if other variables should be used. The current code expects a `gaSettings` variable and this should be removed and possibly replaced with alternative variable expansion.

Yes, `gaSettings` is UA-specific so that wouldn't be usable for GA4, so that part is indeed inaccurate.

Regarding the `constant` variable type, would you please show how can that be used? The docs only mention setting the measurement ID as part of the configuration tag. https://support.google.com/tagmanager/answer/9442095?hl=en

> * The IB specifies changes to [`assets/js/modules/analytics/components/common/ExistingTagNotice.js`](https://github.com/google/site-kit-wp/blob/1735da6226a671f86b1e043796ec4710729e7e06/assets/js/modules/analytics/components/common/ExistingTagNotice.js) but that seems at odds with the fact the current UA message and GTM-related text is in [`assets/js/modules/analytics/components/common/ExistingGTMPropertyNotice.js`](https://github.com/google/site-kit-wp/blob/1735da6226a671f86b1e043796ec4710729e7e06/assets/js/modules/analytics/components/common/ExistingGTMPropertyNotice.js). It looks like it would be more appropriate to add to `ExistingGTMPropertyNotice` and reuse this in the GA4-related flows, i.e. `SetupFormGA4` and `SetupFormGA4Transitional`.

That sounds right to me 👍 

> * The IB mentions showing the same message we currently have when `getAnalyticsPropertyIDs` returns a non-empty array, but all the existing copy is designed with expectation of a single property/measurement ID, the IB doesn't cover what happens when `getAnalyticsPropertyIDs` returns more than one ID.

I'm not sure if any change would be needed in this scenario as the user would be blocked by the current behavior.

---

So yes, it seems that some important details were missing in the IB. Let's bring it back and make sure all these questions are resolved before moving forward.

First, is there anything you feel is still missing from the AC (cc @felixarntz)? If not, would you please update the IB to address the inaccuracies we've identified?

**Comment by eugene-manuilov** [comment]
> > * Additionally the UA dropdown is **not** currently preselected/disabled on the `SetupFormGA4Transitional` flow when a UA tag exists.
> 
> This sounds like it may be a bug although it may have been intentional as a step towards the removal of such restrictions. @eugene-manuilov did we change this recently?

No, we didn't. It should work as far as I remember 🤔

**Comment by techanvil** [comment]
Hi @aaemnnosttv, thanks for the reply and clarification.

> Regarding the constant variable type, would you please show how can that be used? The docs only mention setting the measurement ID as part of the configuration tag. https://support.google.com/tagmanager/answer/9442095?hl=en

I've recorded a clip showing how this can be achieved through the Tag Manager UI. It starts off showing a tag with the measurement ID specified without a variable, then adds a tag which uses a variable of type `constant`.

https://user-images.githubusercontent.com/18395600/145450546-71cbdeec-698e-424b-9a28-a3e89c38fd99.mp4

The subsequent response to a `live-container-version` request is included below, the key difference between the tags being the `measurementId` parameter:

```
        {
          "key": "measurementId",
          "type": "template",
          "value": "G-7TG3MD6C9G"
        }

// vs

        {
          "key": "measurementId",
          "type": "template",
          "value": "{{GA4 Measurement ID}}"
        }

```


<details>
<summary>/wp-json/google-site-kit/v1/modules/tagmanager/data/live-container-version response</summary>

```json
{
  "accountId": "6005158346",
  "containerId": "53575468",
  "containerVersionId": "7",
  "deleted": null,
  "description": null,
  "fingerprint": "1639072553619",
  "name": "add tag with constant",
  "path": "accounts/6005158346/containers/53575468/versions/7",
  "tagManagerUrl": "https://tagmanager.google.com/#/versions/accounts/6005158346/containers/53575468/versions/7?apiLink=version",
  "container": {
    "accountId": "6005158346",
    "containerId": "53575468",
    "domainName": null,
    "fingerprint": "1635175280186",
    "name": "baz.sitekit.10uplabs.com",
    "notes": null,
    "path": "accounts/6005158346/containers/53575468",
    "publicId": "GTM-55CZZTP",
    "tagManagerUrl": "https://tagmanager.google.com/#/container/accounts/6005158346/containers/53575468/workspaces?apiLink=container",
    "usageContext": [
      "web"
    ]
  },
  "tag": [
    {
      "accountId": "6005158346",
      "blockingRuleId": null,
      "blockingTriggerId": null,
      "containerId": "53575468",
      "fingerprint": "1639071902580",
      "firingRuleId": null,
      "firingTriggerId": [
        "2147479553"
      ],
      "liveOnly": null,
      "monitoringMetadataTagNameKey": null,
      "name": "PageView GA4",
      "notes": null,
      "parentFolderId": null,
      "path": null,
      "paused": null,
      "scheduleEndMs": null,
      "scheduleStartMs": null,
      "tagFiringOption": "oncePerEvent",
      "tagId": "3",
      "tagManagerUrl": null,
      "type": "gaawc",
      "workspaceId": null,
      "parameter": [
        {
          "key": "sendPageView",
          "type": "boolean",
          "value": "true"
        },
        {
          "key": "enableSendToServerContainer",
          "type": "boolean",
          "value": "false"
        },
        {
          "key": "measurementId",
          "type": "template",
          "value": "G-7TG3MD6C9G"
        }
      ],
      "monitoringMetadata": {
        "key": null,
        "type": "map",
        "value": null
      },
      "consentSettings": {
        "consentStatus": "notSet"
      }
    },
    {
      "accountId": "6005158346",
      "blockingRuleId": null,
      "blockingTriggerId": null,
      "containerId": "53575468",
      "fingerprint": "1639072541371",
      "firingRuleId": null,
      "firingTriggerId": [
        "2147479553"
      ],
      "liveOnly": null,
      "monitoringMetadataTagNameKey": null,
      "name": "PageView GA4 2",
      "notes": null,
      "parentFolderId": null,
      "path": null,
      "paused": null,
      "scheduleEndMs": null,
      "scheduleStartMs": null,
      "tagFiringOption": "oncePerEvent",
      "tagId": "8",
      "tagManagerUrl": null,
      "type": "gaawc",
      "workspaceId": null,
      "parameter": [
        {
          "key": "sendPageView",
          "type": "boolean",
          "value": "true"
        },
        {
          "key": "enableSendToServerContainer",
          "type": "boolean",
          "value": "false"
        },
        {
          "key": "measurementId",
          "type": "template",
          "value": "{{GA4 Measurement ID}}"
        }
      ],
      "monitoringMetadata": {
        "key": null,
        "type": "map",
        "value": null
      },
      "consentSettings": {
        "consentStatus": "notSet"
      }
    }
  ],
  "variable": [
    {
      "accountId": "6005158346",
      "containerId": "53575468",
      "disablingTriggerId": null,
      "enablingTriggerId": null,
      "fingerprint": "1639072529551",
      "name": "GA4 Measurement ID",
      "notes": null,
      "parentFolderId": null,
      "path": null,
      "scheduleEndMs": null,
      "scheduleStartMs": null,
      "tagManagerUrl": null,
      "type": "c",
      "variableId": "7",
      "workspaceId": null,
      "parameter": [
        {
          "key": "value",
          "type": "template",
          "value": "G-Z8R3SVBJJE"
        }
      ],
      "formatValue": {
        "caseConversionType": null
      }
    }
  ],
  "builtInVariable": [
    {
      "accountId": "6005158346",
      "containerId": "53575468",
      "name": "Page URL",
      "path": null,
      "type": "pageUrl",
      "workspaceId": null
    },
    {
      "accountId": "6005158346",
      "containerId": "53575468",
      "name": "Page Hostname",
      "path": null,
      "type": "pageHostname",
      "workspaceId": null
    },
    {
      "accountId": "6005158346",
      "containerId": "53575468",
      "name": "Page Path",
      "path": null,
      "type": "pagePath",
      "workspaceId": null
    },
    {
      "accountId": "6005158346",
      "containerId": "53575468",
      "name": "Referrer",
      "path": null,
      "type": "referrer",
      "workspaceId": null
    },
    {
      "accountId": "6005158346",
      "containerId": "53575468",
      "name": "Event",
      "path": null,
      "type": "event",
      "workspaceId": null
    }
  ]
}
```

</details>

**Comment by aaemnnosttv** [comment]
Thanks @techanvil – I wasn't aware that this type of variable could be used here but you're right. I think for our users this would likely be a rare edge case but it probably wouldn't be too hard to accommodate it since the infrastructure already exists to look up the variable by name like we have for UA. If the GA4 measurement ID was referenced by a variable, we could get it by name and then check each `parameter.value` for one that was a valid measurement ID (although it looks like a constant only has a single `parameter.value`, it could be generalized in this way). Again, I'm not really sure if we need to account for this since it isn't documented, so it's likely only GTM power users would know how to set this up anyways which is not our average user. Thanks for raising this though.

Thoughts @felixarntz ?

**Comment by techanvil** [comment]
> Thanks @techanvil – I wasn't aware that this type of variable could be used here but you're right. I think for our users this would likely be a rare edge case but it probably wouldn't be too hard to accommodate it since the infrastructure already exists to look up the variable by name like we have for UA. If the GA4 measurement ID was referenced by a variable, we could get it by name and then check each `parameter.value` for one that was a valid measurement ID (although it looks like a constant only has a single `parameter.value`, it could be generalized in this way). Again, I'm not really sure if we need to account for this since it isn't documented, so it's likely only GTM power users would know how to set this up anyways which is not our average user. Thanks for raising this though.
> 
> Thoughts @felixarntz ?

Thanks @aaemnnosttv. I would say that although it's not documented, it's pretty discoverable via the UI, seeing as the button next to the measurement ID input takes the user straight to the variables...

**Comment by aaemnnosttv** [comment]
> I would say that although it's not documented, it's pretty discoverable via the UI, seeing as the button next to the measurement ID input takes the user straight to the variables...

@techanvil true! All I'm saying is that variables are kind of a more advanced feature of GTM but also there is no GA4 specific variable type, so again I think that falls more into a power-user use case to know that a `constant` type variable could even be used there.

**Comment by felixarntz** [comment]
@aaemnnosttv 
> If the GA4 measurement ID was referenced by a variable, we could get it by name and then check each `parameter.value` for one that was a valid measurement ID (although it looks like a constant only has a single `parameter.value`, it could be generalized in this way). Again, I'm not really sure if we need to account for this since it isn't documented, so it's likely only GTM power users would know how to set this up anyways which is not our average user.

Don't we already do the same thing for a UA property ID though? As far as I remember, we allow for the same flexibility there, allow for a hard-coded property ID or a variable reference, and if it's the latter we fetch the variables to find that actual value. I'd say if we can use the same approach for the GA4 measurement ID, we should probably cover that here as well.

cc @techanvil

**Comment by techanvil** [comment]
Hey @felixarntz, just a heads up there are a few more points that could also do with your attention, as per this comment - https://github.com/google/site-kit-wp/issues/4103#issuecomment-989724143

**Comment by aaemnnosttv** [comment]
@felixarntz yes we do, but there are different types of variables which changes the structure of the data a bit and where the relevant value is – it's not a simple key-value. For UA we actually assume a variable there references a `gas` (GA Settings) type variable and look for its value in the relevant item in its list of `parameter` objects.

https://github.com/google/site-kit-wp/blob/59bb881a73e50a2df4805d3de983e48293c86bf9/assets/js/modules/tagmanager/datastore/versions.js#L201-L217

So technically the `constant` variable type that @techanvil mentioned could actually be used for UA too, but it probably wouldn't make much sense as there is already a dedicated variable type for (UA) GA settings.

**Comment by techanvil** [comment]
Hi @aaemnnosttv @felixarntz, thanks for your insight on this ticket. I've updated the IB to address the inaccuracies/details that we've discussed. Sending back over to IBR for a once-over.

**Comment by aaemnnosttv** [comment]
Thanks @techanvil! The IB looks very thorough. Just one comment from me.

> * Using `assets/js/modules/analytics/components/common/GA4ActivateSwitch.js`, don't display the GA4 toggle if the conditions in the previous bullet point are met.

This component would only be shown if GA4 wasn't set up yet in which case there wouldn't be an existing measurement ID set to conflict with one found in a GTM container. The user shouldn't be blocked from setting up GA4, only from double tagging the same property/measurement ID via GA and GTM so in this case, it should still be enabled but the snippet would be disabled the same as if it was already active.

**Comment by techanvil** [comment]
> Thanks @techanvil! The IB looks very thorough. Just one comment from me.
> 
> > * Using `assets/js/modules/analytics/components/common/GA4ActivateSwitch.js`, don't display the GA4 toggle if the conditions in the previous bullet point are met.
> 
> This component would only be shown if GA4 wasn't set up yet in which case there wouldn't be an existing measurement ID set to conflict with one found in a GTM container. The user shouldn't be blocked from setting up GA4, only from double tagging the same property/measurement ID via GA and GTM so in this case, it should still be enabled but the snippet would be disabled the same as if it was already active.

Hi @aaemnnosttv, thanks for highlighting this. I've been trying to determine what this will entail for the implementation. This has led me to notice the `canUseSnippet` code that we've got in place; `canUseSnippet` appears to be false when there is a single GTM-supplied UA tag, true otherwise.

`canUseSnippet` is used to disable the switch in `UseSnippetSwitch` (the UA version):
https://github.com/google/site-kit-wp/blob/037120d66e456e274b1677527295bd08a3789ea1/assets/js/modules/analytics/components/common/UseSnippetSwitch.js#L61-L79

`canUseSnippet` is a setting that appears to be derived from the `gaPropertyID` and `useSnippet` settings.

https://github.com/google/site-kit-wp/blob/037120d66e456e274b1677527295bd08a3789ea1/includes/Modules/Tag_Manager.php#L636-L643

The `gaPropertyID` setting is set from the `getSingleAnalyticsPropertyID` selector.
https://github.com/google/site-kit-wp/blob/037120d66e456e274b1677527295bd08a3789ea1/assets/js/modules/tagmanager/hooks/useGAPropertyIDEffect.js#L32-L40

Note, we haven't specced out a similar `getSingleGA4PropertyID` selector on the GA4 side. Nor is there an equivalent `ga4PropertyID`, and, of course, no GA4 version of `canUseSnippet`.

For reference, here's the rendered `Switch` in the GA4 `UseSnippetSwitch` - note the lack of a `disabled` prop.
https://github.com/google/site-kit-wp/blob/037120d66e456e274b1677527295bd08a3789ea1/assets/js/modules/analytics-4/components/common/UseSnippetSwitch.js#L58-L66

So, I am therefore wondering if we need to replicate this `canUseSnippet` behaviour on the GA4 side. Also, if we are to do that, we will presumably need to set it slightly differently, as we aren't distinguishing the *single* property ID case via a `getSingleGA4PropertyID` selector.

Do you think this `canUseSnippet` approach looks like the way to go here?

**Comment by techanvil** [comment]
This ticket will need to be reassessed once the "Existing tag simplification" epic is complete.

**Comment by aaemnnosttv** [comment]
Thanks @techanvil – as a santity check for myself: @felixarntz do you think this issue should be blocked by the ETS epic or maybe added to it? We definitely don't want to do this at the same time as it would probably result in conflicts, but it could maybe happen before since this is mostly defined.

**Comment by felixarntz** [comment]
@aaemnnosttv Hmm, thinking that it may still be simpler to do it after ETS? It's not urgent so I think it's okay to mark it as blocked by that epic.

**Comment by JayHoltslander** [comment]
Following this as this [recently affected](https://github.com/google/site-kit-wp/issues/5085) one of the sites I look after.

**Comment by FlicHollis** [comment]
Hi @felixarntz @aaemnnosttv @techanvil are you happy to unblock now that we have launched ETS?

**Comment by techanvil** [comment]
@FlicHollis this can be unblocked, we'll need to revisit it to check if it's still occurring, and ensure the AC and IB are up-to-date if so.

**Comment by cooperfellows** [comment]
I can confirm several of our sites are impacted by this as well. Latest version of WordPress and SiteKit installed.

We do make use of the Constant variable type to hold our Measurement ID as was being discussed in December by aaemnnosttv and techanvil. (Intentionally untagged, as our work around is to simply not connect SiteKit to Google Analytics at all until this can resolved, and instead place our GA4 and UA tags via GTM).

**Comment by jamesozzie** [comment]
Tested today and SK does not recognize an existing GA4 snippet if inserted via Tag Manager (manually or via the GTM module within Site Kit). The issue therefore remains but it's no longer specific to a manually placed GTM container. Site Kit does not mention that there is an existing tag recording on their site (via GTM). 

I also think it's safe to close this issue as we have a more recent issue open regarding the same behavior (https://github.com/google/site-kit-wp/issues/7335). After some testing on this today I may expand on that issue, given that there is also a possibility of associating a Google tag with a GA4 measurement ID from within Tag Manager. We therefore need to test out the different scenarios. I'll create a follow up task and update that issue according. 

![image](https://github.com/google/site-kit-wp/assets/41326532/658b5abb-1e6b-421a-9017-d4b98b9a521a)
